backends/edk2-compat/edk2-svc-generate.c

```diff
@@ -139,44 +53,103 @@ int performGenerateCommand(int argc,char* argv[])
 	size_t outBuffSize, size;
 	struct hash_funct *hashFunction;
 	unsigned char *buff = NULL, *outBuff = NULL;
+	char *programName = NULL;
 	struct Arguments args = {	
 		.helpFlag = 0, .inpValid = 0, .signKeyCount = 0, .signCertCount = 0, .alreadySignedFlag = 2,
 		.inFile = NULL, .outFile = NULL,  
 		.signCerts = NULL, .signKeys = NULL, .inForm = NULL, .outForm = NULL, .varName = NULL, 
 		.hashAlg = NULL, .time = NULL
 	};
+	//go back two arguments to get program name and subcommand
+	argc += 2;
+	argv -= 2;
+	//we could iterate to just the subcommand but then the usage would look a bit strange
+	//argp will only allow one string in argv as the program name, we want it to be 'secvarctl generate' 
+	programName = malloc(strlen(argv[0]) + strlen(argv[1]) + strlen(" ") + 1);
+	if (!programName) { 
+		prlog(PR_ERR, "ERROR: failed to allocate memory\n");
+		return ALLOC_FAIL;
+	}
+	programName = strcpy(programName, argv[0]);
+	programName = strcat(programName, " ");
+	programName = strcat(programName, argv[1]);
+	argc--;
+	argv++;
+	argv[0] = programName;
+
```

Comment:
So you're saying have the help message for the specific secvarctl command only display the subcommand (not the `secvarctl`  part). Similar to the output of something like `openssl x509 --help` , where the usage is listed as `Usage: x509 [options]` . I am fine with doing this, as long as everyone thinks it maintains clarity. Please confirm and I can make updates.

Reply:
Sorry, I wasn't clear.

I think it would be fine to print only the subcommand, but I think what you're doing now is also fine. It's not worth getting hung up on; we have bigger fish to fry.

What I want to do in the future is have argument parsing go in 2 stages.

Stage 1 is parsing global options. So here we parse everything up to the subcommand. Usually that will be nothing, but it might be something like `--verbose` or `--help`. Stage 2 is parsing the subcommand arguments, that will specialise by subcommand.

Whether or not that actually makes this particular piece of code simpler in the end I'm not sure, I just wanted to flag this as something I'd want to change later.

Reply:
Just to throw my thoughts in here, I definitely think there should be a two-stage argparse. Along with `--verbose`, `--help`, I think the backend should also be selectable prior to a subcommand, allowing for something along the lines of:
```bash
 # generates for the default/autodetect backend setting
$ secvarctl generate ...  

# generates for the edk2-compat backend explicitly
$ secvarctl --backend edk2-compat generate ...
```

Any command should probably not have to look backwards in argv. Not super familiar with argp, but can we just hardcode the program name arg if we already know what it is supposed to be? (That is, we can infer from where it is being called -- in a backend-specific command)

Reply:
I agree. Perhaps add argp parsing to the secvarctl command when we are actually supporting  multiple backends? It would look a little weird now since the only presubcommand options are `--usage/help` and `-v`.

Also, yes I can hardcode it. I will do that now.

Reply:
I'm happy with this for now, let's revisit it later.